test/e2e/apps/daemon_set.go

```diff
@@ -746,3 +780,189 @@ func waitFailedDaemonPodDeleted(c clientset.Interface, pod *v1.Pod) func() (bool
 		return false, nil
 	}
 }
+
+// deleteDaemonSetAndOrphanPods deletes the given DaemonSet and orphans all its dependents.
+// It also checks that all dependents are orphaned, and the DaemonSet is deleted.
+func deleteDaemonSetAndOrphanPods(
+	dsClient appstyped.DaemonSetInterface,
+	podClient corev1typed.PodInterface,
+	controllerRevisionClient appstyped.ControllerRevisionInterface,
+	ds *apps.DaemonSet) error {
+
+	ds, err := dsClient.Get(ds.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get daemonset %q: %v", ds.Name, err)
+	}
+
+	deletePropagationOrphanPolicy := metav1.DeletePropagationOrphan
+	deleteOptions := &metav1.DeleteOptions{
+		PropagationPolicy: &deletePropagationOrphanPolicy,
+		Preconditions:     metav1.NewUIDPreconditions(string(ds.UID)),
+	}
+	if err = dsClient.Delete(ds.Name, deleteOptions); err != nil {
+		return fmt.Errorf("failed to delete daemonset %q: %v", ds.Name, err)
+	}
+
+	if err = checkDaemonSetDeleted(dsClient, ds.Name); err != nil {
+		return err
+	}
+
+	if err = checkDaemonSetPodsOrphaned(podClient, ds.Name); err != nil {
+		return err
+	}
+
+	if err = checkDaemonSetHistoriesOrphaned(controllerRevisionClient, ds.Name); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkDaemonSetDeleted(dsClient appstyped.DaemonSetInterface, dsName string) error {
+	if err := wait.PollImmediate(dsRetryPeriod, dsRetryTimeout, func() (bool, error) {
+		_, err := dsClient.Get(dsName, metav1.GetOptions{})
+		if !apierrs.IsNotFound(err) {
+			return false, err
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("daemonset %q does not get deleted: %v", dsName, err)
+	}
+	return nil
+}
+
+func getPods(podClient corev1typed.PodInterface, labelMap map[string]string) (*v1.PodList, error) {
+	podSelector := labels.Set(labelMap).AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		return nil, fmt.Errorf("failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	return pods, nil
+}
+
+func checkDaemonSetPodsOrphaned(podClient corev1typed.PodInterface, dsName string) error {
+	label := map[string]string{daemonsetNameLabel: dsName}
+	if err := wait.PollImmediate(dsRetryPeriod, dsRetryTimeout, func() (bool, error) {
+		pods, err := getPods(podClient, label)
+		if err != nil {
+			return false, err
+		}
+
+		for _, pod := range pods.Items {
+			// This pod is orphaned only when its controllerRef is nil
+			if controllerRef := metav1.GetControllerOf(&pod); controllerRef != nil {
+				return false, nil
+			}
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("daemonset pods are not orphaned: %v", err)
```

Comment:
Same comment  here: it could be the list pods failed, but the pods are actually orphaned

Reply:
and everywhere else for most of the wait loop

Reply:
`checkDaemonSetPodsOrphaned` -> `waitDaemonSetPodsOrphaned`